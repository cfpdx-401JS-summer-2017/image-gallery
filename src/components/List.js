import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class List extends PureComponent {
  static propTypes = {
    hondas: PropTypes.arrayOf(PropTypes.object).isRequired,
    deleteImage: PropTypes.func
  };

  render() {
    const { hondas } = this.props;
    return (
      <div className="listView">
        <ul>
          {hondas.map((honda, i) =>
            <li key={i}>
            <span className="title">{honda.title}</span> <span>{honda.description}</span> <span role="img" aria-label=""><a href={i}>&#128279;</a></span>
            </li>
          )}
        </ul>
      </div>
    );
  }
}
