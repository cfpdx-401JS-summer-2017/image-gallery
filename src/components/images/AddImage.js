import React from 'react';
import PropTypes from 'prop-types';

AddImage.propTypes = {
  addImage: PropTypes.func
};

export default function AddImage({ addImage }) {
  console.log('in AddImage: ', addImage)
  return (
    <div className="addImageForm">
      <form
        onSubmit={event => {
          event.preventDefault();
          const form = event.target;
          const { title, desc, url } = form.elements;
          addImage({title:title.value, desc:desc.value, url:url.value });
          form.reset();
        }}>
        <button type="submit" name="add image">
          add image
        </button>
        <input
          type="text"
          className="addImageTextInput"
          name="title"
          placeholder="enter a title for your image"
        />
        <input
          type="text"
          className="addImageTextInput"
          name="desc"
          placeholder="enter a description for your image"
        />
        <input
          type="text"
          className="addImageTextInput"
          name="url"
          placeholder="paste in a new url to add an image"
        />
      </form>
    </div>
  );
}
