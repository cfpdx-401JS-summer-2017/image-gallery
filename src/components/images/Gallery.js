import React, { Component } from 'react';
import PropTypes from 'prop-types';

GalleryItem.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    bunny: PropTypes.object.isRequired,
    onRemove: PropTypes.func.isRequired
}

export function GalleryItem({ title, description, url, bunny, onRemove, onUpdate }) {
    return (
        <div className="galleryView">
            <img src={url} alt={description} />
            <h3>{title}</h3>
            <p>{description}</p>
            <button onClick={() => onRemove(bunny)} >X</button>
        </div>  
    );
}

export class Gallery extends Component {
    
    static propTypes = {
        galleryArray: PropTypes.arrayOf(PropTypes.object).isRequired,
        onRemove: PropTypes.func.isRequired
    }
    
    render() {
        const { galleryArray, onRemove, onUpdate, bunnyNum } = this.props;
        const bunnyGallery = galleryArray.map((bunny, i) => (
            <GalleryItem key={i} bunny={bunny} title={bunny.title} description={bunny.description} url={bunny.url} onRemove={onRemove} onUpdate={onUpdate} />
        ));
        return (
            <div>
                <nav id="galleryNav">
                    <button onClick={() => onUpdate(-1)}>&laquo; Previous</button> <button onClick={() => onUpdate(1)}>Next &raquo;</button>
                    <p id="bunnyCounter">Bunny {bunnyNum + 1} of {bunnyGallery.length}</p>
                </nav>
                {bunnyGallery[bunnyNum]}
            </div>
        );
    }
}