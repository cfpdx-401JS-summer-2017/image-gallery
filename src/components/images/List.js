import React, { Component } from 'react';
import PropTypes from 'prop-types';

ListItem.propTypes = {
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    bunny: PropTypes.object.isRequired,
    onRemove: PropTypes.func.isRequired
}

export function ListItem({ title, description, url, bunny, onRemove }) {
    return (
        <ul>
            <li><h3>{title}</h3></li>
            <li>{description}</li>
            <li><a target="_blank" href={url}>{url}</a></li>
            <button onClick={() => onRemove(bunny)} >X</button>
        </ul>
    );
}

export class List extends Component {

    static propTypes = {
        listArray: PropTypes.arrayOf(PropTypes.object).isRequired,
        onRemove: PropTypes.func.isRequired
    }

    render() {
        const { listArray, onRemove } = this.props;
        return (
            <div>
                {listArray.map((bunny, i) => (
                    <ListItem key={i} bunny={bunny} title={bunny.title} description={bunny.description} url={bunny.url} onRemove={onRemove} />
                ))}
            </div>
        );
    }
}