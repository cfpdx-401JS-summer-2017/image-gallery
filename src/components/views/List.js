import React from 'react';
import PropTypes from 'prop-types';

List.propTypes = {
  hondas: PropTypes.arrayOf(PropTypes.object).isRequired,
  deleteImage: PropTypes.func
};

export default function List({ hondas, deleteImages }) {
  return (
    <div className="listView">
      <ul>
        {hondas.map((honda, i) => (
          <li key={i}>
            <span className="title">{honda.title}</span>
            <span>{honda.description}</span>
            <span role="img" aria-label="">
              <a href={`/images?detail/${i}`}>&#128279;</a>
            </span>
          </li>
        ))}
      </ul>
    </div>
  );
}
