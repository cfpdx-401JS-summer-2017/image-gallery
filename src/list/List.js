import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class List extends Component {

	static propTypes = {
        bunnies: PropTypes.arrayOf(PropTypes.object).isRequired
    }

	constructor(props) {
		super(props);
	}



	render () {
		
	
		return (
			<ul className="list">
				{ this.props.bunnies.map( (bunny) => {
					return (
						<li key={bunny._id}>
						<a href={bunny.url}>{bunny.title}</a>
						<p>{bunny.description}</p>
						</li>)
				})}
			</ul>
		)
	}

}