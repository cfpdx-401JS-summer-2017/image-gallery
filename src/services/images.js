
export function addImage(images, img) {
    let title = img.title;
    let description = img.description;
    let url = img.url;

    const image = { id: ++id, title, url, description }
    return [
        ...images,
        image
    ]
}

export function removeImage(images, image) {
    const index = images.indexOf(image);
    if (index === -1) return images;

    return [
        ...images.slice(0, index),
        ...images.slice(index + 1)
    ]
}

let id = 0;


export const images = [
    { id: ++id, title: 'cute bunny', description: 'very cute bunny', url: 'http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg' },
    { id: ++id, title: 'another cute bunny', description: 'very, very cute bunny', url: 'http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg' },
    { id: ++id, title: 'fairly cute bunny', description: 'pretty much a cute bunny', url: 'http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg' },
    { id: ++id, title: 'ugly bunny', description: 'the ugliest bunny you\'ve ever seen', url: 'https://s-media-cache-ak0.pinimg.com/736x/5c/40/1f/5c401f8c1d1075cc5b49db6cda84700b--kungfu-baby-wombat.jpg' },
];

