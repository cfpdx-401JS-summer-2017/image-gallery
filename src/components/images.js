import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ViewOptions from './images/views/options';
import EmojiList from './images/views/list';
import EmojiThumbnail from './images/views/thumbnail';
import EmojiGallery from './images/views/gallery';

import CreateEmoji from './images/forms/create';
import DeleteEmoji from './images/forms/delete';
import { createEmoji, deleteEmoji } from './images/services/emoji';
import nextId from './images/services/nextId';

const views = {
  list: EmojiList,
  thumbnail: EmojiThumbnail,
  gallery: EmojiGallery,
};

export default class Images extends Component {

  constructor(props) {
    super(props);

    this.state = {
      view: 'list', // key of the views dictionary obj
      emojis: [
        {
          id: nextId(),
          title: 'ship it squirrel',
          description: 'GitHub\'s :shipit: emoji is a mobster squirrel used, typically in PR\'s, to communicate the code is finished and ready for production. Ship it!',
          url: 'https://qph.ec.quoracdn.net/main-qimg-c8781a4bb1f17e330b50cb35f851da05-c'
        },{
          id: nextId(),
          title: 'hamburger',
          description: 'A hamburger or burger is a sandwich consisting of one or more cooked patties of ground meat, usually beef, placed inside a sliced bread roll or bun. The patty may be pan fried, barbecued, or flame broiled. Hamburgers are often served with cheese, lettuce, tomato, bacon, onion, pickles, or chiles; condiments such as mustard, mayonnaise, ketchup, relish, or "special sauce"; and are frequently placed on sesame seed buns. A hamburger topped with cheese is called a cheeseburger.',
          url: 'https://emojipedia-us.s3.amazonaws.com/thumbs/120/apple/96/hamburger_1f354.png',
        },{
          id: nextId(),
          title: 'unicorn',
          description: 'The unicorn is a legendary creature that has been described since antiquity as a beast with a single large, pointed, spiraling horn projecting from its forehead. The unicorn was depicted in ancient seals of the Indus Valley Civilization and was mentioned by the ancient Greeks in accounts of natural history by various writers, including Ctesias, Strabo, Pliny the Younger, and Aelian. The Bible also describes an animal, the re\'em, which some versions translate as unicorn.',
          url: 'https://emojipedia-us.s3.amazonaws.com/thumbs/120/apple/96/unicorn-face_1f984.png',
        },{
          id: nextId(),
          title: 'fire',
          description: 'Fire is the rapid oxidation of a material in the exothermic chemical process of combustion, releasing heat, light, and various reaction products. Slower oxidative processes like rusting or digestion are not included by this definition.',
          url: 'https://emojipedia-us.s3.amazonaws.com/thumbs/120/apple/96/fire_1f525.png'
        },{
          id: nextId(),
          title: 'alien',
          description: 'Extraterrestrial life, also called alien life (or, if it is a sentient or relatively complex individual, an "extraterrestrial" or "alien"), is life that does not originate from Earth. These hypothetical life forms may range from simple single-celled organisms to beings with civilizations far more advanced than humanity. Although many scientists expect extraterrestrial life to exist in some form, there is no evidence for its existence to date. The Drake equation speculates about the existence of intelligent life elsewhere in the universe. The science of extraterrestrial life in all its forms is known as exobiology.',
          url: 'https://emojipedia-us.s3.amazonaws.com/thumbs/120/apple/96/alien-monster_1f47e.png'
        },{
          id: nextId(),
          title: 'nerd',
          description: 'A nerd is a person seen as overly intellectual, obsessive, or lacking social skills. Such a person may spend inordinate amounts of time on unpopular, little known, or non-mainstream activities, which are generally either highly technical, abstract, or relating to topics of fiction or fantasy, to the exclusion of more mainstream activities. Additionally, many so-called nerds are described as being shy, quirky, pedantic, and unattractive, and may have difficulty participating in, or even following, sports.',
          url: 'https://emojipedia-us.s3.amazonaws.com/thumbs/120/apple/96/nerd-face_1f913.png'
        },{
          id: nextId(),
          title: 'floppy disk',
          description: 'A floppy disk, also called a floppy, diskette, or just disk, is a type of disk storage composed of a disk of thin and flexible magnetic storage medium, sealed in a rectangular plastic enclosure lined with fabric that removes dust particles. Floppy disks are read and written by a floppy disk drive (FDD).',
          url: 'https://emojipedia-us.s3.amazonaws.com/thumbs/120/apple/96/floppy-disk_1f4be.png'
        }
      ]
    }
    console.log('starter data', this.state.emojis);

    this.handleViewChange = this.handleViewChange.bind(this);
    this.handleOnSubmitCreateEmoji = this.handleOnSubmitCreateEmoji.bind(this);
    this.handleOnSubmitDeleteEmoji = this.handleOnSubmitDeleteEmoji.bind(this);
  }

  handleViewChange(view) {
    this.setState({ view });
  }

  handleOnSubmitCreateEmoji = (title, description, url) => {
    let emojis = createEmoji(this.state.emojis, title, description, url);
    this.setState({ emojis });
  }
  
  handleOnSubmitDeleteEmoji(id) {
    if (id !== 0) {
      let emojis = deleteEmoji(this.state.emojis, id);
      this.setState({ emojis });
    }
  }

  render() {
    let CurrentView = views[this.state.view];
    
    return (
      <div>
        <ViewOptions onClick={this.handleViewChange} />
        <CurrentView emojis={this.state.emojis} />
        <CreateEmoji handleOnSubmit={this.handleOnSubmitCreateEmoji} />
        <DeleteEmoji emojis={this.state.emojis} handleOnSubmit={this.handleOnSubmitDeleteEmoji} />
      </div>
    );
  }

}

Images.propTypes = {
  view: PropTypes.string,
  emojis: PropTypes.array,
}
