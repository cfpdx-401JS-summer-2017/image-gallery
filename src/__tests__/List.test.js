import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { List, ListItem } from '../components/images/List';

const testFunc = function () { }

describe('ListItem tests', () => {

    it('displays correct properties', () => {
        const bunny = {};
        const wrapper = shallow(<ListItem title="Test Title" description="Test description" url="http://google.com/" bunny={bunny} onRemove={testFunc} />);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

});

describe('List tests', () => {

    it('can render 3 bunnies', () => {

        const wrapper = shallow(

            <List onRemove={testFunc} listArray={[
                {
                    title: 'Cute Bunny 1',
                    description: 'Who\'s a cute bunny??',
                    url: 'http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg',
                    bunny: {},
                    onRemove: testFunc
                },
                {
                    title: 'Cute Bunny 2',
                    description: 'Diabolically cute!!',
                    url: 'http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg',
                    bunny: {},
                    onRemove: testFunc
                },
                {
                    title: 'Cute Bunny 3',
                    description: 'There are no bounds to the cuteness!',
                    url: 'http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg',
                    bunny: {},
                    onRemove: testFunc
                }
            ]} />

        )

        let listArray = wrapper.unrendered.props.listArray;
        expect(listArray.length).toEqual(3);

    });

});
