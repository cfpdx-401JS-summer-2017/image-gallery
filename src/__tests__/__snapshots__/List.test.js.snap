// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem tests displays correct properties 1`] = `
<ul>
  <li>
    <h3>
      Test Title
    </h3>
  </li>
  <li>
    Test description
  </li>
  <li>
    <a
      href="http://google.com/"
      target="_blank"
    >
      http://google.com/
    </a>
  </li>
  <button
    onClick={[Function]}
  >
    X
  </button>
</ul>
`;
